dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -410,20 +430,161 @@ public Workflow refresh(@ApiParam(hidden = true) @Auth User user,
+            @Authorization(value = JWT_SECURITY_DEFINITION_NAME) }, response = Service.class)
+    public Service addService(@ApiParam(hidden = true) @Auth User user,
+            @ApiParam(value = "Repository path", required = true) @PathParam("repository") String repository,
+            @ApiParam(value = "Name of user on GitHub", required = true) @QueryParam("username") String username,
+            @ApiParam(value = "GitHub installation ID", required = true) @QueryParam("installationId") String installationId,
+            @ApiParam(value = "This is here to appease Swagger. It requires PUT methods to have a body, even if it is empty. Please leave it empty.") String emptyBody) {
+        // Check for duplicates (currently workflows and services share paths)
+        String servicePath = "github.com/" + repository;
+        Workflow duplicate = workflowDAO.findByPath(servicePath, false);
 
+        if (duplicate != null) {
+            String msg = "Service with the path " + servicePath + " already exists.";
+            LOG.info(msg);
+            throw new CustomWebApplicationException(msg, HttpStatus.SC_BAD_REQUEST);
+        }
+
+        // Retrieve the user who triggered the call
+        User sendingUser = findUserByGitHubUsername(username);
+
+        // Get Installation Access Token
+        String installationAccessToken = gitHubAppSetup(installationId);
+
+        // Create a service object
+        final SourceCodeRepoInterface sourceCodeRepo = SourceCodeRepoFactory.createGitHubAppRepo(installationAccessToken);
+        Service service = sourceCodeRepo.initializeService(repository);
+        service.getUsers().add(sendingUser);
+        long serviceId = workflowDAO.create(service);
+
+        return (Service)workflowDAO.findById(serviceId);
+    }
+
+    @PUT
+    @Path("/path/service/{repository}/upsertVersion/")
+    @Timed
+    @UnitOfWork
+    @RolesAllowed({ "curator", "admin" })
+    @ApiOperation(value = "Add or update a service version for a given GitHub tag for a service with the given repository (ex. dockstore/dockstore-ui2).", notes = "To be called by a lambda function.", authorizations = {
+            @Authorization(value = JWT_SECURITY_DEFINITION_NAME) }, response = Service.class)
+    public Service upsertServiceVersion(@ApiParam(hidden = true) @Auth User user,
+            @ApiParam(value = "Repository path", required = true) @PathParam("repository") String repository,
+            @ApiParam(value = "Git reference for new GitHub tag", required = true) @QueryParam("gitReference") String gitReference,
+            @ApiParam(value = "GitHub installation ID", required = true) @QueryParam("installationId") String installationId,
+            @ApiParam(value = "This is here to appease Swagger. It requires PUT methods to have a body, even if it is empty. Please leave it empty.") String emptyBody) {
```

Comment:
If this and the previous method were POST instead of PUT, then you wouldn't run into this. Nor would you have the username nor installation id in the url (although I guess neither of those needs to be secured).

Just food for thought; I know it makes things a bit trickier code-wise.

Reply:
Just to be safe I'll convert to post.